### PR TITLE
Cleanup handling of .opt and .exe

### PR DESCRIPTION
**Pull request by nojb**
Following discussion on Slack:

- Only add .opt implicitly for compiler tools
- Do not add .exe implicitly on Windows if already present on the program name

### CODE REVIEW

File: src/stdune/bin.ml [hunk 1]
```diff
@@ -24,24 +24,21 @@ let exists fn =
   | exception (Unix.Unix_error _) -> false
   | _ -> true
 
-let best_prog dir prog =
-  let fn = Path.relative dir (prog ^ ".opt" ^ exe) in
-  if exists fn then
-    Some fn
+let add_exe prog =
+  if String.is_suffix prog ~suffix:exe then
```
Comment (nojb) [inline src/stdune/bin.ml]: I guess I need a case-insensitive check here...

File: src/dune/context.ml [hunk 2]
```diff
@@ -147,8 +147,29 @@ let opam_config_var ~env ~cache var =
...
+
+let which ~path prog =
+  let rec search = function
+    | [] -> None
+    | dir :: rest ->
+      match best_prog dir prog with
+      | None -> search rest
+      | Some fn -> Some fn
+  in
+  search path
```
Comment (nojb) [inline src/dune/context.ml]: I copied this code from `Bin`, but I guess I can use `List.find_map path ~f:(fun dir -> best_prog dir prog)` instead.

Comment (diml) [reply src/dune/context.ml]: Using `List.find_map` seems good indeed

File: src/stdune/bin.ml (hunk 1 above)
Comment (diml) [reply src/stdune/bin.ml]: Indeed

File: src/stdune/bin.ml [hunk 3]
```diff
@@ -24,24 +24,21 @@ let exists fn =
...
+    prog ^ exe
 
 let which ~path prog =
   let rec search = function
     | [] -> None
     | dir :: rest ->
-      match best_prog dir prog with
-      | None -> search rest
-      | Some fn -> Some fn
+      let fn = Path.relative dir (add_exe prog) in
```
Comment (diml) [inline src/stdune/bin.ml]: We can do the `add_exe` just once outside the loop

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/stdune/bin.ml (hunk 1 above)
Comment (nojb) [reply src/stdune/bin.ml]: Done.

File: src/dune/context.ml (hunk 2 above)
Comment (nojb) [reply src/dune/context.ml]: Done.

File: src/stdune/bin.ml (hunk 3 above)
Comment (nojb) [reply src/stdune/bin.ml]: Indeed. Fixed.

**Review by ghost — APPROVED** [review]
Looks good! This need a changelog entry BTW

**Comment by nojb** [comment]
> Looks good! This need a changelog entry BTW

Added. I will merge once CI passes. Thanks!

File: src/dune/context.ml [hunk 4]
```diff
@@ -151,7 +151,20 @@ let opam_config_var ~env ~cache var =
         | Error _ ->
             None ) )
 
-let which ~cache ~path x = Table.find_or_add cache x ~f:(Bin.which ~path)
+let best_prog dir prog =
+  let fn = Path.relative dir (prog ^ ".opt" ^ Bin.exe) in
+  if Bin.exists fn then
+    Some fn
+  else
+    let fn = Path.relative dir (prog ^ Bin.exe) in
```
Comment (rgrinberg) [inline src/dune/context.ml]: Option.some_if makes this a bit cleaner.

File: src/stdune/bin.ml [hunk 5]
```diff
@@ -31,25 +31,20 @@ let exists fn =
...
-    | [] ->
-        None
-    | dir :: rest -> (
-      match best_prog dir prog with None -> search rest | Some fn -> Some fn )
-  in
-  search path
+  let prog = add_exe prog in
+  List.find_map path ~f:(fun dir ->
+      let fn = Path.relative dir prog in
+      if exists fn then
```
Comment (rgrinberg) [inline src/stdune/bin.ml]: Ditto

Comment (nojb) [reply src/stdune/bin.ml]: I agree, done.

File: src/dune/context.ml (hunk 4 above)
Comment (nojb) [reply src/dune/context.ml]: Here I'm not convinced; if I use `Option.some_if` around the first call to `Bin.exists` then I need to pattern match on the result, which is not nice. And if I only use it around the second call to `Bin.exists` then the code loses its symmetry.

Comment (rgrinberg) [reply src/dune/context.ml]: The symmetry was lost on me :) anyway it's just a suggestion, feel free to leave as is.

**Pull request merged**